File: tom_som_cpp_runtime/tests/facade_editability_test.cpp
/* Behavioural test for a generated root facade's static `editabilityFor`
 * (§ item 8). The facade cannot be regenerated here (the generator runs
 * centrally, once, for all languages), so this hand-writes the exact shape the
 * `som_cpp_emitter` now emits per root — a static `editabilityFor` delegating to
 * `som::somEditabilityFor(kModelVersion, documentVersion)` — and asserts it
 * classifies correctly. This pins the emitter contract independently of a full
 * regeneration.
 */
#include <cstdio>
#include <string>

#include "som_facade.hpp"

namespace {

// Mirror of the generated root facade's editability surface (kModelVersion +
// the static editabilityFor the emitter now produces). kModelVersion "1.2" is a
// stand-in for the model version stamped into the real generated root.
struct GeneratedRootFacade {
  static constexpr const char* kModelVersion = "1.2";
  static som::SomEditability editabilityFor(const std::string& documentVersion) {
    return som::somEditabilityFor(kModelVersion, documentVersion);
  }
};

std::size_t g_passed = 0;
std::size_t g_failed = 0;

void check(const char* name, bool cond) {
  if (cond) {
    g_passed++;
  } else {
    g_failed++;
    std::printf("  FAIL: %s\n", name);
  }
}

}  // namespace

int main() {
  using E = som::SomEditability;

  check("root editabilityFor empty -> editable",
        GeneratedRootFacade::editabilityFor("") == E::editable);
  check("root editabilityFor older minor -> editable",
        GeneratedRootFacade::editabilityFor("1.0") == E::editable);
  check("root editabilityFor equal -> editable",
        GeneratedRootFacade::editabilityFor("1.2") == E::editable);
  check("root editabilityFor newer minor -> rejected",
        GeneratedRootFacade::editabilityFor("1.5") == E::rejectedNewerMinor);
  check("root editabilityFor cross-major -> read-only",
        GeneratedRootFacade::editabilityFor("2.0") == E::readOnlyCrossMajor);
  check("root editabilityFor invalid -> invalidVersion",
        GeneratedRootFacade::editabilityFor("nope") == E::invalidVersion);

  if (g_failed == 0) {
    std::printf("facade_editability_test: all %zu checks passed\n", g_passed);
    return 0;
  }
  std::printf("facade_editability_test: %zu passed, %zu FAILED\n", g_passed,
              g_failed);
  return 1;
}

File: tom_som_go_runtime/tests/facade_editability_test.go
// Behavioural test for the root facade's generated EditabilityFor method
// (SOM §21).
//
// The generated facade (tom_som_go_v0) cannot be regenerated here (that must run
// centrally, once, for all nine languages), so this test hand-writes a minimal
// root facade that mirrors exactly what SomGoEmitter now emits per root:
//
//	const fakeRootModelVersion = "1.2"
//	func (x *fakeRoot) EditabilityFor(documentVersion string) som.SomEditability {
//	    return som.SomEditabilityFor(fakeRootModelVersion, documentVersion)
//	}
//
// It asserts the delegating method returns the same classification as the
// underlying SomEditabilityFor for every SOM §4.2 case.
package tests

import (
	"testing"

	som "github.com/al-the-bear/tom_ai_build/tom_som_go_runtime"
)

// fakeRootModelVersion stands in for the per-root <Root>ModelVersion const the
// emitter allocates.
const fakeRootModelVersion = "1.2"

// fakeRoot is a stand-in for a generated root facade struct (embeds SomNode,
// carries the generated EditabilityFor method delegating to the runtime).
type fakeRoot struct {
	som.SomNode
}

// EditabilityFor mirrors the emitted root method: it delegates to the runtime
// classifier with the root's own model-version const.
func (x *fakeRoot) EditabilityFor(documentVersion string) som.SomEditability {
	return som.SomEditabilityFor(fakeRootModelVersion, documentVersion)
}

func TestRootEditabilityFor(t *testing.T) {
	root := &fakeRoot{SomNode: som.NewSomNode(som.NewSpecDocument(), "root")}

	cases := []struct {
		name     string
		document string
		want     som.SomEditability
	}{
		{"empty stamp editable", "", som.SomEditabilityEditable},
		{"same major older minor editable", "1.0", som.SomEditabilityEditable},
		{"same major equal minor editable", "1.2", som.SomEditabilityEditable},
		{"same major newer minor rejected", "1.5", som.SomEditabilityRejectedNewerMinor},
		{"cross major read-only", "2.0", som.SomEditabilityReadOnlyCrossMajor},
		{"invalid stamp", "bogus", som.SomEditabilityInvalidVersion},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := root.EditabilityFor(tc.document)
			if got != tc.want {
				t.Fatalf("root.EditabilityFor(%q) = %d, want %d", tc.document, got, tc.want)
			}
			// The delegating method must agree with the underlying classifier.
			if got != som.SomEditabilityFor(fakeRootModelVersion, tc.document) {
				t.Fatalf("root.EditabilityFor(%q) diverged from SomEditabilityFor", tc.document)
			}
		})
	}
}
